### ISSUE TITLE
[FEATURE] Follow official Go guidance on how to organize a module?

### ISSUE THREAD

**Issue by craigpastro** [open-feature/go-sdk#227]
### Requirements

Go released an official guidance on how to organize a module here: https://go.dev/doc/modules/layout. Do you think it is a good idea to follow this guidance for this repo? What it means here is to just move everything in `pkg/openfeature` to the top-level. IMHO it would make importing this module a bit nicer:
```
import "github.com/open-feature/go-sdk/pkg/openfeature"
```
vs
```
import "github.com/open-feature/go-sdk"
```

The mismatch of the package name and the import is a bit jarring though. I suppose one could also "import as".

Unfortunately, this would be a breaking change.

**Comment by toddbaert** [comment]
@Kavindu-Dodan @thisthat @bacherfl @RealAnna do you folks have any opinions on this?

**Comment by paddycarver** [comment]
`github.com/open-feature/go-sdk/openfeature` would be a nice import path, because it helps `goimports` to automatically import the package when it's used in a file.

You can probably do this without a breaking change using [type aliases](https://go.dev/doc/go1.9#language) to make the types of the `pkg/openfeature` package aliases of the types in the `openfeature` package, and having functions in the `pkg/openfeature` package call their identically-named functions in the `openfeature` directory. Constants can likewise be aliases into the the `openfeature` package.

The `memprovider` package could take a similar approach.

I think that would preserve all identifiers and behaviors, while only maintaining a single implementation of the behaviors. Future identifiers wouldn't need to be added to `pkg/openfeature`. And because the types are considered the same types, there shouldn't be any compilation impact.

**Comment by beeme1mr** [comment]
I think this is worth exploring, especially if we could make this a non-breaking change. @paddycarver, would you be willing to put together a proof of concept using type aliases?

**Comment by paddycarver** [comment]
I've opened #232 and #233 with proofs of concept of the github.com/open-feature/go-sdk/openfeature and github.com/open-feature/go-sdk import paths, respectively. I don't have time to do the harder work (validation), but the branches build and the tests pass. I don't have a consumer of the package at-hand to verify the compatibility, unfortunately.

**Comment by Kavindu-Dodan** [comment]
Thank you @paddycarver for the POC PRs. I checked them and I'm in favor of the new packaging proposed through https://github.com/open-feature/go-sdk/pull/232  (`root/pkg/openfeature` -> `root/openfeature`). I think this packaging is ideal, solves the packaging issue as well as isolates logic instead of having it at root level (i.e- compared to @233)

I tried adapting go sdk providers and all looks fine. However, this make lint failures which means we have to update all references to the latest packaging. 

@toddbaert  shall we proceed with this change?

**Comment by beeme1mr** [comment]
I think this is a great idea. Thanks @paddycarver, for putting together the PoC!

@craigpastro are you happy with this solution?

**Comment by toddbaert** [comment]
> Thank you @paddycarver for the POC PRs. I checked them and I'm in favor of the new packaging proposed through #232 (`root/pkg/openfeature` -> `root/openfeature`). I think this packaging is ideal, solves the packaging issue as well as isolates logic instead of having it at root level (i.e- compared to #233)
> 
> I tried adapting go sdk providers and all looks fine. However, this make lint failures which means we have to update all references to the latest packaging.
> 
> @toddbaert shall we proceed with this change?

I don't have a problem with correcting the lint errors. I think this is a very worthwhile change.

I'm very happy with this solution.

**Comment by craigpastro** [comment]
I think this is a great solution that avoids breakage. Thanks @paddycarver 🎉

**Comment by Kavindu-Dodan** [comment]
Created https://github.com/open-feature/go-sdk/issues/234 to track follow-up tasks once this issue is resolved with the proposed fix at #232
